### PR TITLE
[Log Analytics] Fix help + examples

### PR DESCRIPTION
**Pull request by alexeldeib**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `./scripts/ci/test_static.sh` locally? (`pip install pylint flake8` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### CODE REVIEW

**Comment by alexeldeib** [comment]
@adewaleo Think it should be good. I bumped the package version and link in index. 

image

**Comment by alexeldeib** [comment]
@mmacy FYI, and thanks for pointing this out.

File: src/log-analytics/azext_loganalytics/_help.py [hunk 1]
```diff
@@ -7,22 +7,29 @@
...
-        - analytics-query: --analytics-query
-          type: string
-          short-summary: Query to execute over the Log Analytics data.
-        - timespan: --timespan -t
-          type: string
-          short-summary: Timespan over which to query data. Defaults to all data.
-        - workspaces: --workspaces
-          type: array
-          short-summary: Additional workspaces to union data for querying. Specify additional workspace IDs separated by commas.
+      - name: --workspace
```
Comment (adewaleo) [inline src/log-analytics/azext_loganalytics/_help.py]: No need to add help for the parameters here as they will be gotten from [_params.py](https://github.com/Azure/azure-cli-extensions/blob/90a71cac2d7b6915fa5d73d8f2d7e11699e6c9d7/src/log-analytics/azext_loganalytics/_params.py). It can be beneficial to add help for parameters here if you need to add more content like [long summaries or populator-commands](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_help.md).  

You don't need to remove the help you added here, just let them be consistent with the help in your _params.py file.

File: src/log-analytics/azext_loganalytics/_help.py [hunk 2]
```diff
@@ -7,22 +7,29 @@
...
+        type: string
+        short-summary: GUID of the Log Analytics workspace.
+      - name: --analytics-query
+        type: string
+        short-summary: Query to execute over the Log Analytics data.
+      - name: --timespan
```
Comment (adewaleo) [inline src/log-analytics/azext_loganalytics/_help.py]: I believe the CI test you are failing is because you don't have the correct parameter help names for some of your parameters. In[ _params.py](https://github.com/Azure/azure-cli-extensions/blob/90a71cac2d7b6915fa5d73d8f2d7e11699e6c9d7/src/log-analytics/azext_loganalytics/_params.py), you have:

- your `workspace` parameter has the options `options_list=['--workspace', '-w']`. However, in your help.py your parameter name is `--workspace` instead of `--workspace -w`
- your `timespan` parameter has the options `options_list=['--timespan', '-t']`. However, in your help.py your parameter name is `--timespan` instead of `--timespan -t`

So to solve this problem, you can: 
- use the old help.py yaml parameter names before your change
- or remove the help entry for your **all** your parameters, so the help can be generated from _params.py.

Comment (alexeldeib) [reply src/log-analytics/azext_loganalytics/_help.py]: The old parameter names don't seem to parse any more, seems like some validation changed on the key fields?

Additionally, I [tried using the naming](https://github.com/Azure/azure-cli-extensions/pull/316/commits/65753c4ea35fd5c90585f5555a2b2fbc6d3bde04) you recommend after finding [this issue[(https://github.com/Microsoft/knack/issues/85). It didn't work for me, the validation still failed CI. 

Conveniently, if I remove the parameters which fail name validation from the help file, they autopopulate the parameter help correctly. The remaining parameters pass name validation but the help generation doesn't always work, so I added a string in the help file to populate the --analytics-query parameter properly.

File: src/log-analytics/azext_loganalytics/_help.py (hunk 1 above)
Comment (alexeldeib) [reply src/log-analytics/azext_loganalytics/_help.py]: This did not work 100% for me. --analytics-query did not load any text from the parameter python file (or elsewhere). Did work for other parameters, though.

**Comment by alexeldeib** [comment]
@adewaleo CI seems happy now 😄 Barring any more feedback, this looks good to me.

File: src/log-analytics/HISTORY.rst [hunk 3]
```diff
@@ -1,12 +1,22 @@
+0.1.4
+++++++++++++++++++
```
Comment (adewaleo) [inline src/log-analytics/HISTORY.rst]: Shouldn't there be only one history.rst change?

File: src/log-analytics/azext_loganalytics/_help.py [hunk 4]
```diff
@@ -7,22 +7,20 @@
...
-        - analytics-query: --analytics-query
-          type: string
-          short-summary: Query to execute over the Log Analytics data.
-        - timespan: --timespan -t
-          type: string
-          short-summary: Timespan over which to query data. Defaults to all data.
-        - workspaces: --workspaces
-          type: array
-          short-summary: Additional workspaces to union data for querying. Specify additional workspace IDs separated by commas.
+      - name: --analytics-query
```
Comment (adewaleo) [inline src/log-analytics/azext_loganalytics/_help.py]: Is there a particular reason we have this parameter's help here? If you aren't adding any other attributes to this parameter's help like `long-summary` then it's best to remove this parameter's help as well.

File: src/log-analytics/setup.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 from codecs import open
 from setuptools import setup, find_packages
 
-VERSION = "0.1.2"
+VERSION = "0.1.4"
```
Comment (adewaleo) [inline src/log-analytics/setup.py]: Shouldn't this be 0.1.3?

File: src/log-analytics/azext_loganalytics/_help.py (hunk 2 above)
Comment (adewaleo) [reply src/log-analytics/azext_loganalytics/_help.py]: Yes, if help for a parameter is missing in the Yaml file, it would be generated from the help argument in _params.py, as mentioned above.

Hmm that's strange that using the old naming didn't work. 

I am happy though that you were able to resolve this issue.

File: src/log-analytics/HISTORY.rst (hunk 3 above)
Comment (alexeldeib) [reply src/log-analytics/HISTORY.rst]: Yes, PyPi.

Comment (alexeldeib) [reply src/log-analytics/HISTORY.rst]: https://pypi.org/manage/project/log-analytics/releases/ see past releases

File: src/log-analytics/azext_loganalytics/_help.py (hunk 4 above)
Comment (adewaleo) [reply src/log-analytics/azext_loganalytics/_help.py]: Was this before or after adding the help information for this parameter in _params.py. This [line](https://github.com/Azure/azure-cli-extensions/blob/24f3fcd2c3aa81b8a157339fb998fc09c325fec7/src/log-analytics/azext_loganalytics/_params.py#L12) of code in _params.py should provide the help summary. If you add it, you shouldn't need to add this parameter's help information in _help.py. Just make sure the string you pass in for help is the short summary you need.

File: src/log-analytics/HISTORY.rst (hunk 3 above)
Comment (adewaleo) [reply src/log-analytics/HISTORY.rst]: Ah I see. Thanks for explaining. I think next time it is better to have the PR merged in before publishing the whl file on pypi.

File: src/log-analytics/azext_loganalytics/_help.py (hunk 4 above)
Comment (adewaleo) [reply src/log-analytics/azext_loganalytics/_help.py]: Please try this out and let me know if it works.

Comment (alexeldeib) [reply src/log-analytics/azext_loganalytics/_help.py]: Using the following code (snippets): 

```python
def load_arguments(self, _):
    with self.argument_context('monitor log-analytics query') as c:
        c.argument('workspace', options_list=['--workspace', '-w'], help='GUID of the Log Analytics Workspace')
        c.argument('analytics-query', options_list=['--analytics-query'], help='Query to execute over Log Analytics data.')
        c.argument('timespan', options_list=['--timespan', '-t'], help='Timespan over which to query. Defaults to querying all available data.')
        c.argument('workspaces', nargs='+', help='Additional workspaces to union data for querying. Specify additional workspace IDs separated by space.')
```

```python
helps['monitor log-analytics query'] = """
    type: command
    short-summary: Query a Log Analytics workspace.
    examples:
      - name: Execute a simple query over past 3.5 days.
        text: |
          az monitor log-analytics query -w b8317023-66e4-4edc-8a5b-7c002b22f92f --analytics-query "AzureActivity | summarize count() by bin(timestamp, 1h)" -t P3DT12H
"""
```

I see the following results (note the missing text on the first line):

image

Comment (adewaleo) [reply src/log-analytics/azext_loganalytics/_help.py]: Did you rebuild the whl file and re-add your extension after making these changes?

Comment (alexeldeib) [reply src/log-analytics/azext_loganalytics/_help.py]: Yes, I rebuilt the wheel and added the extension. This is not from a previous version of code. I ran this right now. The commands are cut off in the screenshot, this is the exact sequence I ran.

```powershell
rmdir -Force ./dist
python setup.py bdist_wheel
az extension remove -n log-analytics
az extension add -s ./dist/log_analytics-0.1.4-py2.py3-none-any.whl
az monitor log-analytics query -h
```

Comment (alexeldeib) [reply src/log-analytics/azext_loganalytics/_help.py]: (to be clear, I haven't changed the version again locally which is why it matches the one in PyPi)

Comment (adewaleo) [reply src/log-analytics/azext_loganalytics/_help.py]: Okay, this is strange, let me look into this. I will checkout your pr.

Comment (adewaleo) [reply src/log-analytics/azext_loganalytics/_help.py]: @alexeldeib after a while of debugging. I found the issue. 

TL;DR

your destination command parameter, **analytics-query**, name in _params.py does not match your method parameter name, **analytics_query**, in custom.py.

P.s. you do not need `options_list` if you parameter name is the same as your method name with the hyphens swapped for underscores and without the first two hyphens. 

See [here](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#customize-arguments) for info on customizing command arguments.

======================================================

Long explanation:

so if you look in _params.py you have:

```
...
        c.argument('analytics-query', options_list=['--analytics-query'], help='Query to execute over Log Analytics data.')
                          ^
                          |
dest method parameter name

...
```
whereas in custom.py you have:

```
...
def execute_query(client, workspace, analytics_query, timespan=None, workspaces=None):
                                         ^
                                         |
               method parameter name
...
``` 
Do you see the difference between the variable your are registering the parameter with and the variable you are passing to your method?

When customizing your parameter, you are saying `--analytics-query` maps to some method parameter `analytics-query`, that you want to add help text to. However this function parameter does not exist in your method definition. 

On the other hand you did not explicitly customize the `analytics_query` method param. So it has no special help or options. The cli automatically generates a vanilla command parameter of `--analytics-query`. 

By the way if your destination method parameter name is correct, you do not need to add an option list. The cli would know to convert your method parameter names with underscores to a suitable command parameter name with hyphens.

Comment (alexeldeib) [reply src/log-analytics/azext_loganalytics/_help.py]: Totally missed the hyphen vs. underscore, thank you for helping with the debug! And my apologies for sending you on a lengthy debug. I can make the code change and bump version, or leave as is? Alternatively, I can move storage from PyPi to an Azure blob and stick with 0.1.3 version as it *should* be.

Comment (adewaleo) [reply src/log-analytics/azext_loganalytics/_help.py]: Oh no worries. I missed it too, lol. If you can move your storage to an azure blob, that would be nice. Would the whl file being on pypi affect users of your extension? Would you be able to remove or update the whl file on pypi so as not to confuse your users?

Comment (alexeldeib) [reply src/log-analytics/azext_loganalytics/_help.py]: After I upload to an Azure blob, I'll delete the PyPi project. Should have zero impact.
